internal/zinc-classfile/src/main/scala/sbt/internal/inc/STJ.scala

```diff
@@ -0,0 +1,306 @@
+   * system path. A class in a jar is identified as a path to the jar
+   * and path to the class within that jar (`RelClass`). Those two values
+   * are held in one string separated by `!`. Slashes in both
+   * paths are consistent with `File.separatorChar` as the actual
+   * string is usually kept in `File` object.
+   *
+   * As an example: given a jar file "C:\develop\zinc\target\output.jar"
+   * and a relative path to the class "sbt/internal/inc/Compile.class"
+   * The resulting identifier would be:
+   * "C:\develop\zinc\target\output.jar!sbt\internal\inc\Compile.class"
+   */
+  class JaredClass(override val toString: String) extends AnyVal {
+
+    def relClass: RelClass = toJarAndRelClass._2
+
+    def toJarAndRelClass: (File, RelClass) = {
+      val Array(jar, cls) = toString.split("!")
+      // JaredClass stores RelClass part with File.separatorChar, however actual paths in zips always use '/'
+      val relClass = cls.replace('\\', '/')
+      (new File(jar), relClass)
+    }
+
+    /**
+     * Wraps the string value inside a [[java.io.File]] object.
+     * File is needed to e.g. be compatible with [[xsbti.compile.analysis.ReadStamps]] interface.
+     */
+    def toFile: File = new File(toString)
+
+  }
+
+  object JaredClass {
+
+    /**
+     * The base constructor for `JaredClass`
+     * @param jar the jar file
+     * @param cls the relative path to class within the jar
+     * @return a proper JaredClass identified by given jar and path to class
+     */
+    def apply(jar: File, cls: RelClass): JaredClass = {
+      // This identifier will be stored as a java.io.File. Its constructor will normalize slashes
+      // which means that the identifier to be consistent should at all points have consistent
+      // slashes for safe comparisons, especially in sets or maps.
+      val relClass = if (File.separatorChar == '/') cls else cls.replace('/', File.separatorChar)
+      new JaredClass(s"$jar!$relClass")
+    }
+
+    // Converts URL to JaredClass but reuses the jar file extracted at the call site to avoid recomputing.
+    /**
+     * Converts an URL to a class in a jar to `JaredClass`. The method is rather trivial
+     * as it also takes precomputed path to the jar that it logically should extract itself.
+     * However as it is computed at the callsite anyway, to avoid recomputation it is passed
+     * as a parameter/
+     *
+     * As an example, given a URL:
+     * "jar:file:///C:/develop/zinc/target/output.jar!/sbt/internal/inc/Compile.class"
+     * and a file: "C:\develop\zinc\target\output.jar"
+     * it will create a `JaredClass` represented as:
+     * "C:\develop\zinc\target\output.jar!sbt\internal\inc\Compile.class"
+     *
+     * @param url url to a class inside a jar
+     * @param jar a jar file where the class is located in
+     * @return the class inside a jar represented as `JaredClass`
+     */
+    def fromURL(url: URL, jar: File): JaredClass = {
```

Comment:
This is an odd method... it feels like if the caller has already split the URL to find the file, re-splitting it here isn't necessary.

Reply:
I agree that it is odd. Take a look at sbt.internal.inc.classfile.Analyze#urlAsFile. It previously was simply `IO.urlAsFile` and it was enough. When compiling to plain files, classloader would either return a plain file or url to a class in jar. `IO.urlAsFile` would just ignore the class file and take the jar. But now we can get url to file in jar and we want to keep the class part as long as the jar is the output jar that we are compiling to. So I reuse `IO.urlAsFile` to convert url to file properly and then, if the extracted file is our output jar I know that I need to create a `JaredClass`. To do this, I extract the relative path to class and combine it with already extracted file to avoid doing it twice. I just did not want to copy code from `sbt.io.IO`, especially that it handles some odd corner cases. I'd feel more comfortable that it was in one place. What do you think about this?